From 42c966eb410ee9fb0427703e3499a8e867f53636 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
Date: Thu, 4 Jul 2024 13:37:23 +0800
Subject: [PATCH 107/149] mtk: hostapd: Fix the op_class value in the RNR IE to
 achieve backward compatibility

According to the chapter 11.49 Reduced neighbor report
of DraftP802.11be_D5.1, the AP should select (one of) the operating
class(es) that is expected to be understood by all STAs that might
intend to connect to the reported AP, even if the channel spacing of
that operating class is less than the BSS bandwidth of all the reported
APs in the Neighbor AP Information field.

For example, if an AP reports an EHT AP that is operating a BSS with 320
MHz bandwidth in 6 GHz band using operating class 137, it is recommended
that the operating class selected by the AP in the Reduced Neighbor Report
element is operating class 131, 132, 133 or 134, since these are expected
to be understood by all STAs that can connect to the reported 6GHz AP.

Signed-off-by: Evelyn Tsai <evelyn.tsai@mediatek.com>
Signed-off-by: Money Wang <money.wang@mediatek.com>
Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 src/ap/ieee802_11.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 7762f88f4..ab25d8f53 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7975,7 +7975,7 @@ static bool hostapd_eid_rnr_bss(struct hostapd_data *hapd,
 		 * one TBTT info available. */
 		*tbtt_count_pos = eid++;
 		*eid++ = tbtt_info_len;
-		*eid++ = op_class;
+		*eid++ = (op_class == 137 ? 134 : op_class);
 		*eid++ = bss->iconf->channel;
 		*len += RNR_TBTT_HEADER_LEN;
 	}
-- 
2.45.2

